Bluetooth: host: Fix compilation issue when privacy is disabled

joerchan opened the pull request:
Fix compilation issue when extended advertising is enabled but privacy
is disabled. In this case the rpa_update work is compiled out.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Fix compilation issue when privacy is disabled

Fix compilation issue when extended advertising is enabled but privacy
is disabled. In this case the rpa_update work is compiled out.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented:
To prevent this from happening again could you also enable some sample or test in CI for the configuration combination that was failing?

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1254,6 +1254,19 @@ static inline bool rpa_timeout_valid_check(void)
+	/* RPA is considered new if there is less than half a second since the
+	 * timeout was started.
+	 */
+	return k_delayed_work_remaining_get(&bt_dev.rpa_update) >
+	       (RPA_TIMEOUT - K_MSEC(500));
```
This will need to be fixed almost as soon as it's merged, since with latest upstream when `CONFIG_LEGACY_TIMEOUT_API` is not enabled both `RPA_TIMEOUT` and `K_MSEC()` are structs whereas the return value of `k_delayed_work_remaining_get()` is `u32_t`. In fact, why not make this immediately the "right way" so that it works with the new `k_timeout_t`-based world?

jhedberg replied:
From what I can see both RPA_TIMEOUT and K_MSEC(500) could just be kept in milliseconds, which would solve this.

jhedberg replied:
..that said I guess modifying RPA_TIMEOUT type has a trickle effect, so I'm starting to wonder if a follow-up PR wouldn't make more sense. I'll let you choose :)

joerchan replied:
@jhedberg I think the best approach is to do this in a follow-up PR after this and https://github.com/zephyrproject-rtos/zephyr/pull/23911.

joerchan committed:
Bluetooth: tests: Add additional build configurations

Add additional build configurations to the shell to catch build errors
when enabling extended advertising.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
> To prevent this from happening again could you also enable some sample or test in CI for the configuration combination that was failing?

I have added 2 configurations to the shell as build only targets. I think we should wait until we have extended advertising support in the controller before adding samples.

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1130,6 +1130,19 @@ static int set_le_scan_enable(u8_t enable)
 }
 #endif /* CONFIG_BT_OBSERVER */
 
+static inline bool rpa_is_new(void)
```
Note: Having this function as inline avoids compilation warnings for "defined but not used".
I think this is the best way to handle this conditional struct variables, and we can add more of these for e.g. bt_dev.rl_entries. That should enable us to clean up a lot so that we can use IS_ENABLED much more.

jhedberg commented:
@joerchan note the Shippable failure. You should use %zd for ssize_t and %zu for size_t. Actually, I don't see anything like that in the code you're changing/adding, so seems like you unveiled pre-existing issues in the code which you'll now have to fix in this PR :)

joerchan committed:
Bluetooth: Use zd print conversion modifier for ssize_t

Use the zd print conversion modifier for ssizet_t variables.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
> @joerchan note the Shippable failure. You should use %zd for ssize_t and %zu for size_t. Actually, I don't see anything like that in the code you're changing/adding, so seems like you unveiled pre-existing issues in the code which you'll now have to fix in this PR :)

Thanks. It must be because I enabled the tests using native_posix board. In sanity-check this builds with  `-m64`, while locally for me it builds with ` -m32`. I don't know why that happens though.

joerchan commented on subsys/bluetooth/host/settings.c:
```diff
@@ -137,7 +137,7 @@ static int set(const char *name, size_t len_rd, settings_read_cb read_cb,
 		if (len < sizeof(bt_dev.id_addr[0])) {
```
@jhedberg While going over ssize_t and zd change I noticed this comparison. This is ssize_t vs size_t comparison.
If I enable `-Wsign-compare` I get a worning on this line (and similar for IRK).
Should we fix this one as well?

carlescufi replied:
I would definitely fix it

joerchan replied:
Noted. Will not fix as part of this PR though.

carlescufi reviewed (approved):
looks good, thanks!
